docs/contributing.rst

```diff
@@ -54,103 +54,58 @@ where appropriate.
-  versions.
-
-- ``tox -e cover`` checks the test coverage only. Calling the
-  ``./tox.cover.sh`` script directly (or even ``./tox.cover.sh $pkg1
-  $pkg2 ...`` for any subpackages) might be a bit quicker, though.
-
-- ``tox -e lint`` checks the style of the whole project, while
-  ``pylint --rcfile=.pylintrc path`` will check a single file or
-  specific directory only.
-
-- For debugging, we recommend ``pip install ipdb`` and putting
-  ``import ipdb; ipdb.set_trace()`` statement inside the source
-  code. Alternatively, you can use Python's standard library `pdb`,
-  but you won't get TAB completion...
-
+When you are working in a file foo.py, there should also be a file foo_test.py
+(if there isn't, make one). While you are working on your code and tests, run
+``python foo_test.py`` to run the relevant tests.
+
+For debugging, we recommend ``pip install ipdb`` and putting
+``import ipdb; ipdb.set_trace()`` statement inside the source
+code. Alternatively, you can use Python's standard library `pdb`,
+but you won't get TAB completion.
+
+Once you are done with your code changes, and the tests in foo_test.py pass,
+run all of the unittests for Certbot with ``tox -e py27`` (this uses Python
+2.7).
+
+Once all the unittests pass, check for sufficient test coverage using
+``tox -e cover``, and then check for code style with ``tox -e lint`` (all files)
+or ``pylint --rcfile=.pylintrc path/to/file.py`` (single file at a time).
+
+Once all of the above is successful, you may run the full test suite,
+including integration tests, using ``tox``. We recommend running the
+commands above first, because running all tests with ``tox`` is very
+slow, and the large amount of ``tox`` output can make it hard to find
+specific failures when they happen. Also note that the full test suite
+will attempt to modify your system's Apache config if your user has sudo
+permissions, so it should not be run on a production Apache server.
+
+If you have trouble getting the full ``tox`` suite to run locally, it is
+generally sufficient to open a pull request and let Github and Travis run
+integration tests for you.
 
 .. _integration:
 
-Integration testing with the boulder CA
+Integration testing with the Boulder CA
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Generally it is sufficient to open a pull request and let Github and Travis run
-integration tests for you.
-
-However, if you prefer to run tests, you can use Vagrant, using the Vagrantfile
-in Certbot's repository. To execute the tests on a Vagrant box, the only
-command you are required to run is::
-
-  ./tests/boulder-integration.sh
-
-Otherwise, please follow the following instructions.
+To run integration tests locally, you need Docker and docker-compose installed
+and working. Fetch and start Boulder using:
 
-macOS users: Run ``./tests/mac-bootstrap.sh`` instead of
-``boulder-start.sh`` to install dependencies, configure the
-environment, and start boulder.
+  ./tests/boulder-fetch.sh
```

Comment:
Sorry, I should have said in the PR description: These test instructions assume https://github.com/certbot/certbot/pull/4180 is landed.